File: cpp/se/StandardObjectType.cpp
#include "se/StandardObjectType.hpp"

using namespace std;

vector<const EnumType *> StandardObjectType::values;
vector<string> StandardObjectType::stdURIs;

const StandardObjectType* StandardObjectType::DOMAIN()
{
	static StandardObjectType obj(
		"domain",
		"urn:ietf:params:xml:ns:domain-1.0",
		"urn:ietf:params:xml:ns:domain-1.0 domain-1.0.xsd",
		"name");
	return &obj;
}

const StandardObjectType* StandardObjectType::HOST()
{
	static StandardObjectType obj(
		"host",
		"urn:ietf:params:xml:ns:host-1.0",
		"urn:ietf:params:xml:ns:host-1.0 host-1.0.xsd",
		"name");
	return &obj;
}

const StandardObjectType* StandardObjectType::CONTACT()
{
	static StandardObjectType obj(
		"contact",
		"urn:ietf:params:xml:ns:contact-1.0",
		"urn:ietf:params:xml:ns:contact-1.0 contact-1.0.xsd",
		"id");
	return &obj;
}

StandardObjectType::StandardObjectType(const string& name,
                                       const string& uri,
                                       const string& schemaLocation,
                                       const string& identType)
    : EnumType(values, name, uri, schemaLocation, identType)
{ }

const StandardObjectType* StandardObjectType::value(const string &name)
{
    return (const StandardObjectType *)EnumType::value (name, values);
}


const vector<string>& StandardObjectType::getStandardURIs()
{
    if (stdURIs.size() == 0)
    {
        stdURIs.push_back(DOMAIN()->getURI());
        stdURIs.push_back(HOST()->getURI());
        stdURIs.push_back(CONTACT()->getURI());
    }
    
    return stdURIs;
}

void StandardObjectType::init()
{
    getStandardURIs();
	DOMAIN();
	HOST();
	CONTACT();
}

File: src/main/java/com/ausregistry/jtoolkit2/se/StandardObjectType.java
package com.ausregistry.jtoolkit2.se;

/**
 * This defines the enumeration that maps a EPP namespace alias to a namespace definition and the identifier tag in the
 * resulting XML.
 */
public enum StandardObjectType implements ObjectType {
    HOST("host", "urn:ietf:params:xml:ns:host-1.0", "host-1.0.xsd", "name"),
    CONTACT("contact", "urn:ietf:params:xml:ns:contact-1.0", "contact-1.0.xsd", "id"),
    DOMAIN("domain", "urn:ietf:params:xml:ns:domain-1.0", "domain-1.0.xsd", "name");

    private static String[] stdURIs;
    private String name, uri, schemaDefinition, identType;

    static {
        stdURIs = new String[] {
            DOMAIN.getURI(),
            HOST.getURI(),
            CONTACT.getURI()
        };
    }

    StandardObjectType(final String name, final String uri, final String schemaDefinition,
            final String identType) {

        this.name = name;
        this.uri = uri;
        this.schemaDefinition = schemaDefinition;
        this.identType = identType;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getURI() {
        return uri;
    }

    @Override
    public String getSchemaLocation() {
        return uri + " " + schemaDefinition;
    }

    public String getSchemaDefintion() {
        return schemaDefinition;
    }

    @Override
    public String getIdentType() {
        return identType;
    }

    public static String[] getStandardURIs() {
        return stdURIs;
    }
}

